### ISSUE TITLE
code design recommendations

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#202]
The more I use the library, the more problems I run into. A thing that has currently no good solution for me is to how to deal with deeply nested element hierarchies where I need to access member functions.

I started with how examples are written, that is, many factory-like functions that produce hierarchies. This works well but problems begin when I need to access member functions.

- `share()`ing these elements and using them from elsewhere is possible but hardly maintainable. I can not upcast the pointer because I lose the access to member functions. But I just can not put `std::shared_ptr<proxy</* 10+ instantiations */>>` everywhere in other code.
- To avoid passing very long type names around the code, I create a ton of lambdas and pass function objects instead. This is doable, but my user-state-types are being flooded with function objects and writing these functions when creating UI hierarchies means I mix my logic/state code with the UI which doesn't appeal to me. Also, the lambdas capture a lot (often 4+ elements).
- Some of my data requires custom rendering. I made custom elements from them, implemening the element interface. This works well, but I don't think it would be good to use this pattern for every situation as more code wants to manage other elements than to perform custom drawing.

**Comment by djowel** [comment]
- Use shared_ptr<T> for known primitives (leaf types whenever possible)
- Use shared_ptr<decltype(some-expression)> if you want to capture more complex types. For example:

 ```c++
using check_box_type = decltype(elements::check_box(""));
using toggle_button_ptr = std::shared_ptr<check_box_type>;
```
- Your last item: custom elements

**Comment by Xeverous** [comment]
So in short, you say that using shared pointers and/or custom elements is the cleanest solution?

**Comment by djowel** [comment]
> So in short, you say that using shared pointers and/or custom elements is the cleanest solution?

I think so, yes. The other option is what most OOP-style libs do: capture an abstract base class or mixin class and do the dynamic/polymorphic cast to the derived as needed. I'd prefer no casts, IMO. Sometimes it is unavoidable when static and dynamic types are both used (as done in elements). It's a design issue that boils down to one question: how do you store a complex type? You either 1. erase the type (and do type recovery later) or 2. keep the full static type and deal with the messy complex type.

**Comment by Xeverous** [comment]
After some discussions [elsewhere](https://www.reddit.com/r/learnprogramming/comments/i9jgj2/design_pattern_for_implementation_of_dependent/), I came up with a solution that satisfies me and solves 2 problems (type management and callback hell):

- Instead of one giant file with all the UI elements in it (and very sphagetti lambdas that bind to various structs), I separated certain parts (subhierarchies) into separate classes. Each of these classes:
    - stores user state with public getters, this data updates on UI changes
    - stores shared pointers to elements which it manages
    - has a function that returns `std::shared_ptr<element>` to it's root element. This erases the type, but beyond the scope of this class code is not really interested what it is exactly, outside code just `hold()`s the result.
    - gets a write-only access to global event queue
- I hooked my event queue processing into `view::poll`, so it runs my state machine on every frame.

The result is that each class has good encapsulation and preserves own invariants and arbitrary event can cause arbitrary action to be taken without strong coupling or cyclic dependencies within the code.
